gsy_framework/sim_results/electric_blue/volume_timeseries.py

```diff
@@ -0,0 +1,158 @@
+            energy_kWh, asset_stats.time_slot, market_type)
+        for time_slot, value in profile.items():
+            self._add_to_volume_time_series(time_slot, value, market_type, "bought_kWh")
+
+    def _add_total_energy_sold(
+            self, asset_stats: ForwardDeviceStats, market_type: AvailableMarketTypes):
+        """Add statistics for the total amount of sold energy."""
+        energy_kWh = asset_stats.total_energy_sold
+        if energy_kWh <= 0:
+            return
+        profile = ForwardTradeProfileGenerator(self.asset_capacity).generate_trade_profile(
+            energy_kWh, asset_stats.time_slot, market_type
+        )
+        for time_slot, value in profile.items():
+            self._add_to_volume_time_series(time_slot, value, market_type, "sold_kWh")
+
+    def _add_to_volume_time_series(
+            self, time_slot: DateTime, value: float,
+            market_type: AvailableMarketTypes, attribute_name: str):
+        """Add time slot value to the correct time slot to asset volume time series."""
+        time_slot = self._adapt_time_slot(time_slot)
+        volume_time_series = self._get_asset_volume_time_series(time_slot.start_of("year"))
+        volume_time_series[time_slot][str(market_type)][attribute_name] += value
+
+    def _adapt_time_slot(self, time_slot: DateTime) -> DateTime:
+        """Find the containing time slot of the smaller time slot."""
+        if self.resolution == AggregationResolution.RES_15_MINUTES:
+            return time_slot.set(minute=(time_slot.minute // 15) * 15)
+        return time_slot.start_of(START_OF[self.resolution])
+
+    def _get_asset_volume_time_series(self, time_slot):
+        """Return asset volume time series for the required time slot.
+        If not found in the buffer, it tries fetching it from DB.
+        If not found in the DB, it will generate a new one for the whole year."""
+        if time_slot in self._asset_volume_time_series_buffer:
+            time_series = self._asset_volume_time_series_buffer[time_slot]
+        else:
+            time_series = self._fetch_asset_volume_time_series_from_db(time_slot)
+            if time_series is None:
+                time_series = {
+                    ts: self._get_time_series_template(value)
+                    for ts, value in self._generate_asset_volume_time_series(time_slot)}
+            self._asset_volume_time_series_buffer[time_slot] = time_series
+        return time_series
+
+    def _fetch_asset_volume_time_series_from_db(self, time_slot: DateTime):
+        # TODO: should be implemented by Kamil.
+        return None
+
+    def _generate_asset_volume_time_series(self, time_slot: DateTime):
+        """Generate asset volume time series for the whole year."""
+        start_time = self._adapt_time_slot(time_slot)
+        if start_time < time_slot:
+            start_time += self.resolution.duration()
+
+        end_time = self._adapt_time_slot(time_slot.add(years=1))
```

Comment:
The time series will always contain a complete year (from it's beginning to it's end). So, if you start from 1st of March this year unitl 1st of March next year, you will get time series for 2 complete years!

Now, if it's the first time, the whole time series will be generated. Otherwise it will be fetched from DB.